src/Microsoft.AspNetCore.Mvc.Core/Internal/FileContentResultExecutor.cs

```diff
@@ -15,15 +22,110 @@ public FileContentResultExecutor(ILoggerFactory loggerFactory)
+            {
+                return Task.CompletedTask;
+            }
+
+            else
+            {
+                try
+                {
+                    return response.Body.WriteAsync(result.FileContents, offset: (int)range.From.Value, count: (int)rangeLength);
+                }
+
+                catch (OperationCanceledException ex)
+                {
+                    // Don't throw this exception, it's most likely caused by the client disconnecting.
+                    // However, if it was cancelled for any other reason we need to prevent empty responses.
+                    context.HttpContext.Abort();
+                    return Task.FromException(ex);
+                }
+
+            }
         }
 
-        private static Task WriteFileAsync(ActionContext context, FileContentResult result)
+        private RangeItemHeaderValue SetContentRangeAndStatusCode(ActionContext context, FileContentResult result)
         {
+            var ranges = ParseRange(context, result);
             var response = context.HttpContext.Response;
+            var httpResponseHeaders = response.GetTypedHeaders();
+            bool rangeNotSatisfiable = false;
+            if (ranges == null || ranges.Count != 1)
+            {
+                rangeNotSatisfiable = true;
+            }
+
+            if (rangeNotSatisfiable)
+            {
+                // 14.16 Content-Range - A server sending a response with status code 416 (Requested range not satisfiable)
+                // SHOULD include a Content-Range field with a byte-range-resp-spec of "*". The instance-length specifies
+                // the current length of the selected resource.  e.g. */length
+                httpResponseHeaders.ContentRange = new ContentRangeHeaderValue(result.FileContents.Length);
+                response.StatusCode = StatusCodes.Status416RangeNotSatisfiable;
+                response.ContentLength = result.FileContents.Length;
+                return null;
+            }
+
+            // Multi-range is not supported.
+            Debug.Assert(ranges.Count == 1);
+
+            var range = ranges.SingleOrDefault();
+            httpResponseHeaders.ContentRange = new ContentRangeHeaderValue(
+                range.From.Value,
+                range.To.Value,
+                result.FileContents.Length);
+
+            httpResponseHeaders.LastModified = result.LastModified;
+            httpResponseHeaders.ETag = result.EntityTag;
+            response.StatusCode = StatusCodes.Status206PartialContent;
+            return range;
+        }
+
+        private ICollection<RangeItemHeaderValue> ParseRange(ActionContext context, FileContentResult result)
+        {
+            var httpContext = context.HttpContext;
+            var httpRequestHeaders = httpContext.Request.GetTypedHeaders();
+            var response = httpContext.Response;
+
+            var lastModified = result.LastModified;
+            var etag = result.EntityTag;
```

Comment:
nit: I would just inline these variables given that you are already calling the method below within multiple lines.